crates/sgx_validation/src/lib.rs

```diff
@@ -74,43 +77,65 @@ impl Sgx {
-            }
-
-            if rpt.mrsigner != [0u8; 32] {
-                return Err(anyhow!("untrusted enarx signer"));
-            }
-
-            if rpt.cpusvn != [0u8; 16] {
-                return Err(anyhow!("untrusted cpu"));
-            }
-
-            if rpt.attributes() != Attributes::default() {
-                return Err(anyhow!("untrusted attributes"));
-            }
-
-            if rpt.misc_select() != MiscSelect::default() {
-                return Err(anyhow!("untrusted misc select"));
-            }
-
-            if rpt.enclave_product_id() != u16::MAX {
-                return Err(anyhow!("untrusted enclave product id"));
-            }
-
-            if rpt.enclave_security_version() < u16::MAX {
-                return Err(anyhow!("untrusted enclave"));
+            // Validate that the certification request came from an SGX enclave.
+            let hash = Sha256::digest(cri.public_key.to_vec()?);
+            ensure!(
+                hash.as_slice() == &rpt.reportdata[..hash.as_slice().len()],
+                "sgx report data is invalid"
+            );
+
+            if let Some(config) = config {
+                if !config.measurements.signer.is_empty() {
+                    let signed = config.measurements.signer.contains(&rpt.mrsigner);
+                    ensure!(signed, "sgx untrusted enarx signer");
+                }
+
+                if !config.measurements.hash.is_empty() {
+                    let approved = config.measurements.hash.contains(&rpt.mrenclave);
+                    ensure!(approved, "sgx untrusted enarx hash");
+                }
+
+                if !config.measurements.hash_blacklist.is_empty() {
+                    let denied = config.measurements.hash_blacklist.contains(&rpt.mrenclave);
+                    ensure!(!denied, "sgx untrusted enarx hash");
+                }
+
+                if let Some(product_id) = config.enclave_product_id {
+                    ensure!(
+                        rpt.enclave_product_id() == product_id,
+                        "sgx untrusted enclave product id",
+                    );
+                }
+
+                if let Some(version) = config.enclave_security_version {
+                    ensure!(
+                        rpt.enclave_security_version() >= version,
+                        "sgx untrusted enclave security version"
+                    );
+                }
+
+                if !config.features.is_empty()
+                    && !rpt
+                        .attributes()
+                        .features()
+                        .difference(config.features)
+                        .is_empty()
+                {
+                    bail!("sgx untrusted features");
+                }
+
+                if !config.misc_select.is_empty() {
+                    ensure!(
+                        rpt.misc_select().difference(config.misc_select).is_empty(),
```

Comment:
```suggestion
                        rpt.misc_select().is_disjoint(&config.misc_select),
```

https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.is_disjoint

Reply:
Oh, this is probably not a `HashSet`